From 10f6fd2d90527f78bcf9a70a52d9ac97a589e645 Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Wed, 20 Aug 2025 13:44:47 -0400
Subject: [PATCH] Issue 6947 - Revise time skew check in healthcheck tool and
 add option to exclude checks

Description:

The current check reports a critical warning if time skew is greater than
1 day - even if "nsslapd-ignore-time-skew" is set to "on". If we are ignoring
time skew we should still report a warning if it's very significant like
30 days.

Also added an option to exclude checks

Relates: https://github.com/389ds/389-ds-base/issues/6947

Reviewed by: progier, spichugi, viktor(Thanks!!!)
---
 .../suites/healthcheck/health_config_test.py  |  3 +-
 .../suites/healthcheck/health_repl_test.py    |  1 +
 .../healthcheck/health_security_test.py       |  1 +
 .../healthcheck/health_tunables_test.py       |  1 +
 .../suites/healthcheck/healthcheck_test.py    | 76 ++++++++++++++++---
 src/lib389/lib389/cli_ctl/health.py           | 34 +++++++--
 src/lib389/lib389/dseldif.py                  | 42 +++++++---
 src/lib389/lib389/lint.py                     | 11 +++
 8 files changed, 141 insertions(+), 28 deletions(-)

diff --git a/dirsrvtests/tests/suites/healthcheck/health_config_test.py b/dirsrvtests/tests/suites/healthcheck/health_config_test.py
index 168b5092e..1173aec04 100644
--- a/dirsrvtests/tests/suites/healthcheck/health_config_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/health_config_test.py
@@ -10,7 +10,7 @@
 import pytest
 import os
 import subprocess
-
+import time
 from lib389.backend import Backends, DatabaseConfig
 from lib389.cos import CosTemplates, CosPointerDefinitions
 from lib389.dbgen import dbgen_users
@@ -41,6 +41,7 @@ def run_healthcheck_and_flush_log(topology, instance, searched_code, json, searc
     args.list_checks = False
     args.check = ['config', 'refint', 'backends', 'monitor-disk-space', 'logs', 'memberof']
     args.dry_run = False
+    args.exclude_check = []
 
     # If we are using BDB as a backend, we will get error DSBLE0006 on new versions
     if ds_is_newer("3.0.0") and instance.get_db_lib() == 'bdb' and \
diff --git a/dirsrvtests/tests/suites/healthcheck/health_repl_test.py b/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
index 0c1fa783a..fbd06571b 100644
--- a/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/health_repl_test.py
@@ -322,6 +322,7 @@ def test_healthcheck_non_replicated_suffixes(topology_m2):
     args.check = ['backends']
     args.dry_run = False
     args.json = False
+    args.exclude_check = []
 
     health_check_run(inst, topology_m2.logcap.log, args)
 
diff --git a/dirsrvtests/tests/suites/healthcheck/health_security_test.py b/dirsrvtests/tests/suites/healthcheck/health_security_test.py
index fd6dd8044..283e87995 100644
--- a/dirsrvtests/tests/suites/healthcheck/health_security_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/health_security_test.py
@@ -39,6 +39,7 @@ def run_healthcheck_and_flush_log(topology, instance, searched_code, json, searc
     args.list_checks = False
     args.check = ['config', 'encryption', 'tls', 'fschecks']
     args.dry_run = False
+    args.exclude_check = []
 
     if json:
         log.info('Use healthcheck with --json option')
diff --git a/dirsrvtests/tests/suites/healthcheck/health_tunables_test.py b/dirsrvtests/tests/suites/healthcheck/health_tunables_test.py
index 44d3adf1f..9c6a0ec58 100644
--- a/dirsrvtests/tests/suites/healthcheck/health_tunables_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/health_tunables_test.py
@@ -32,6 +32,7 @@ def run_healthcheck_and_flush_log(topology, instance, searched_code=None, json=F
     args.verbose = instance.verbose
     args.list_errors = list_errors
     args.list_checks = list_checks
+    args.exclude_check = []
     args.check = check
     args.dry_run = False
     args.json = json
diff --git a/dirsrvtests/tests/suites/healthcheck/healthcheck_test.py b/dirsrvtests/tests/suites/healthcheck/healthcheck_test.py
index 7afcd588d..1f09abdde 100644
--- a/dirsrvtests/tests/suites/healthcheck/healthcheck_test.py
+++ b/dirsrvtests/tests/suites/healthcheck/healthcheck_test.py
@@ -30,18 +30,22 @@ ds_paths = Paths()
 log = logging.getLogger(__name__)
 
 
-def run_healthcheck_and_flush_log(logcap, instance, searched_code=None, json=False, searched_code2=None,
-                                  list_checks=False, list_errors=False, check=None, searched_list=None):
+def run_healthcheck_and_flush_log(logcap, instance, searched_code=None,
+                                  json=False, searched_code2=None,
+                                  list_checks=False, list_errors=False,
+                                  check=None, searched_list=None):
     args = FakeArgs()
     args.instance = instance.serverid
     args.verbose = instance.verbose
     args.list_errors = list_errors
     args.list_checks = list_checks
     args.check = check
+    args.exclude_check = []
     args.dry_run = False
     args.json = json
 
-    # If we are using BDB as a backend, we will get error DSBLE0006 on new versions
+    # If we are using BDB as a backend, we will get error DSBLE0006 on new
+    # versions
     if ds_is_newer("3.0.0") and instance.get_db_lib() == 'bdb' and \
        (searched_code is CMD_OUTPUT or searched_code is JSON_OUTPUT):
         searched_code = 'DSBLE0006'
@@ -52,17 +56,37 @@ def run_healthcheck_and_flush_log(logcap, instance, searched_code=None, json=Fal
     if searched_list is not None:
         for item in searched_list:
             assert logcap.contains(item)
-            log.info('Healthcheck returned searched item: %s' % item)
+            log.info('Healthcheck returned searched item: %s', item)
     else:
         assert logcap.contains(searched_code)
-        log.info('Healthcheck returned searched code: %s' % searched_code)
+        log.info('Healthcheck returned searched code: %s', searched_code)
 
     if searched_code2 is not None:
         if ds_is_newer("3.0.0") and instance.get_db_lib() == 'bdb' and \
-        (searched_code2 is CMD_OUTPUT or searched_code2 is JSON_OUTPUT):
+           (searched_code2 is CMD_OUTPUT or searched_code2 is JSON_OUTPUT):
             searched_code = 'DSBLE0006'
         assert logcap.contains(searched_code2)
-        log.info('Healthcheck returned searched code: %s' % searched_code2)
+        log.info('Healthcheck returned searched code: %s', searched_code2)
+
+    log.info('Clear the log')
+    logcap.flush()
+
+
+def run_healthcheck_exclude(logcap, instance, unwanted, wanted, exclude_check):
+    args = FakeArgs()
+    args.instance = instance.serverid
+    args.verbose = instance.verbose
+    args.list_errors = False
+    args.list_checks = False
+    args.check = None
+    args.exclude_check = [exclude_check]
+    args.dry_run = False
+    args.json = False
+
+    health_check_run(instance, logcap.log, args)
+
+    assert not logcap.contains(unwanted)
+    assert logcap.contains(wanted)
 
     log.info('Clear the log')
     logcap.flush()
@@ -127,7 +151,7 @@ def test_healthcheck_standalone(topology_st):
 
     standalone = topology_st.standalone
 
-    run_healthcheck_and_flush_log(topology_st.logcap, standalone, CMD_OUTPUT,json=False)
+    run_healthcheck_and_flush_log(topology_st.logcap, standalone, CMD_OUTPUT, json=False)
     run_healthcheck_and_flush_log(topology_st.logcap, standalone, JSON_OUTPUT, json=True)
 
 
@@ -276,8 +300,40 @@ def test_healthcheck_check_option(topology_st):
         run_healthcheck_and_flush_log(topology_st.logcap, standalone, searched_code=JSON_OUTPUT, json=True, check=[item])
 
 
-@pytest.mark.ds50873
-@pytest.mark.bz1685160
+def test_healthcheck_exclude_option(topology_st):
+    """Check functionality of HealthCheck Tool with --exclude-check option
+
+    :id: a4e2103c-67b8-4359-a8ba-67a8650cd3b7
+    :setup: Standalone instance
+    :steps:
+        1. Set check to exclude from list
+        2. Run HealthCheck
+    :expectedresults:
+        1. Success
+        2. Success
+    """
+
+    inst = topology_st.standalone
+
+    exclude_list = [
+        ('config:passwordscheme', 'config:passwordscheme',
+         'config:securitylog_buffering'),
+        ('config', 'config:', 'backends:userroot:mappingtree')
+    ]
+
+    for exclude, unwanted, wanted in exclude_list:
+        unwanted_pattern = 'Checking ' + unwanted
+        wanted_pattern = 'Checking ' + wanted
+
+        log.info('Exclude check: %s unwanted: %s wanted: %s',
+                 exclude, unwanted, wanted)
+
+        run_healthcheck_exclude(topology_st.logcap, inst,
+                                unwanted=unwanted_pattern,
+                                wanted=wanted_pattern,
+                                exclude_check=exclude)
+
+
 @pytest.mark.skipif(ds_is_older("1.4.1"), reason="Not implemented")
 def test_healthcheck_standalone_tls(topology_st):
     """Check functionality of HealthCheck Tool on TLS enabled standalone instance with no errors
diff --git a/src/lib389/lib389/cli_ctl/health.py b/src/lib389/lib389/cli_ctl/health.py
index d85e3906a..38540e0df 100644
--- a/src/lib389/lib389/cli_ctl/health.py
+++ b/src/lib389/lib389/cli_ctl/health.py
@@ -75,6 +75,9 @@ def _list_errors(log):
 
 
 def _list_checks(inst, specs: Iterable[str]):
+    if specs is None:
+        yield []
+        return
     o_uids = dict(_list_targets(inst))
     for s in specs:
         wanted, rest = DSLint._dslint_parse_spec(s)
@@ -85,19 +88,27 @@ def _list_checks(inst, specs: Iterable[str]):
             for l in o_uids[wanted].lint_list(rest):
                 yield o_uids[wanted], l
         else:
-            raise ValueError('No such object specifier')
+            raise ValueError('No such object specifier: ' + wanted)
 
 
 def _print_checks(inst, log, specs: Iterable[str]) -> None:
     for o, s in _list_checks(inst, specs):
         log.info(f'{o.lint_uid()}:{s[0]}')
 
-def _run(inst, log, args, checks):
+
+def _run(inst, log, args, checks, exclude_checks):
     if not args.json:
         log.info("Beginning lint report, this could take a while ...")
 
     report = []
+    excludes = []
+    for _, skip in exclude_checks:
+        excludes.append(skip[0])
+
     for o, s in checks:
+        if s[0] in excludes:
+            continue
+
         if not args.json:
             log.info(f"Checking {o.lint_uid()}:{s[0]} ...")
         try:
@@ -119,12 +130,12 @@ def _run(inst, log, args, checks):
         if count > 1:
             plural = "s"
         if not args.json:
-            log.info("{} Issue{} found!  Generating report ...".format(count, plural))
+            log.info(f"{count} Issue{plural} found!  Generating report ...")
             idx = 1
             for item in report:
                 _format_check_output(log, item, idx)
                 idx += 1
-            log.info('\n\n===== End Of Report ({} Issue{} found) ====='.format(count, plural))
+            log.info(f'\n\n===== End Of Report ({count} Issue{plural} found) =====')
         else:
             log.info(json.dumps(report, indent=4))
 
@@ -147,17 +158,21 @@ def health_check_run(inst, log, args):
     dsrc_inst = dsrc_to_ldap(DSRC_HOME, args.instance, log.getChild('dsrc'))
     dsrc_inst = dsrc_arg_concat(args, dsrc_inst)
     try:
-        inst = connect_instance(dsrc_inst=dsrc_inst, verbose=args.verbose, args=args)
+        inst = connect_instance(dsrc_inst=dsrc_inst, verbose=args.verbose,
+                                args=args)
     except Exception as e:
-        raise ValueError('Failed to connect to Directory Server instance: ' + str(e))
+        raise ValueError('Failed to connect to Directory Server instance: ' +
+                         str(e)) from e
 
     checks = args.check or dict(_list_targets(inst)).keys()
-
+    exclude_checks = args.exclude_check
+    print("MARK excl: " + str(exclude_checks))
     if args.list_checks or args.dry_run:
         _print_checks(inst, log, checks)
         return
 
-    _run(inst, log, args, _list_checks(inst, checks))
+    _run(inst, log, args, _list_checks(inst, checks),
+         _list_checks(inst, exclude_checks))
 
     disconnect_instance(inst)
 
@@ -175,3 +190,6 @@ def create_parser(subparsers):
     run_healthcheck_parser.add_argument('--check', nargs='+', default=None,
                                         help='Areas to check. These can be obtained by --list-checks. Every element on the left of the colon (:)'
                                              ' may be replaced by an asterisk if multiple options on the right are available.')
+    run_healthcheck_parser.add_argument('--exclude-check', nargs='+', default=[],
+                                        help='Areas to skip. These can be obtained by --list-checks. Every element on the left of the colon (:)'
+                                             ' may be replaced by an asterisk if multiple options on the right are available.')
diff --git a/src/lib389/lib389/dseldif.py b/src/lib389/lib389/dseldif.py
index 20eb0bb52..d12c6424c 100644
--- a/src/lib389/lib389/dseldif.py
+++ b/src/lib389/lib389/dseldif.py
@@ -23,7 +23,8 @@ from lib389.lint import (
     DSPERMLE0002,
     DSSKEWLE0001,
     DSSKEWLE0002,
-    DSSKEWLE0003
+    DSSKEWLE0003,
+    DSSKEWLE0004
 )
 
 
@@ -68,26 +69,49 @@ class DSEldif(DSLint):
         return 'dseldif'
 
     def _lint_nsstate(self):
+        """
+        Check the nsState attribute, which contains the CSN generator time
+        diffs, for excessive replication time skew
+        """
+        ignoring_skew = False
+        skew_high = 86400  # 1 day
+        skew_medium = 43200  # 12 hours
+        skew_low = 21600  # 6 hours
+
+        ignore_skew = self.get("cn=config", "nsslapd-ignore-time-skew")
+        if ignore_skew is not None and ignore_skew[0].lower() == "on":
+            # If we are ignoring time skew only report a warning if the skew
+            # is significant
+            ignoring_skew = True
+            skew_high = 86400 * 365  # Report a warning for skew over a year
+            skew_medium = 99999999999
+            skew_low = 99999999999
+
         suffixes = self.readNsState()
         for suffix in suffixes:
             # Check the local offset first
             report = None
-            skew = int(suffix['time_skew'])
-            if skew >= 86400:
-                # 24 hours - replication will break
-                report = copy.deepcopy(DSSKEWLE0003)
-            elif skew >= 43200:
+            skew = abs(int(suffix['time_skew']))
+            if skew >= skew_high:
+                if ignoring_skew:
+                    # Ignoring skew, but it's too excessive not to report it
+                    report = copy.deepcopy(DSSKEWLE0004)
+                else:
+                    # 24 hours of skew - replication will break
+                    report = copy.deepcopy(DSSKEWLE0003)
+            elif skew >= skew_medium:
                 # 12 hours
                 report = copy.deepcopy(DSSKEWLE0002)
-            elif skew >= 21600:
+            elif skew >= skew_low:
                 # 6 hours
                 report = copy.deepcopy(DSSKEWLE0001)
             if report is not None:
                 report['items'].append(suffix['suffix'])
                 report['items'].append('Time Skew')
                 report['items'].append('Skew: ' + suffix['time_skew_str'])
-                report['fix'] = report['fix'].replace('YOUR_INSTANCE', self._instance.serverid)
-                report['check'] = f'dseldif:nsstate'
+                report['fix'] = report['fix'].replace('YOUR_INSTANCE',
+                                                      self._instance.serverid)
+                report['check'] = 'dseldif:nsstate'
                 yield report
 
     def _update(self):
diff --git a/src/lib389/lib389/lint.py b/src/lib389/lib389/lint.py
index 460bf64fc..2ca39e0d6 100644
--- a/src/lib389/lib389/lint.py
+++ b/src/lib389/lib389/lint.py
@@ -528,6 +528,17 @@ Also look at https://access.redhat.com/documentation/en-us/red_hat_directory_ser
 and find the paragraph "Too much time skew"."""
 }
 
+DSSKEWLE0004 = {
+    'dsle': 'DSSKEWLE0004',
+    'severity': 'Low',
+    'description': 'Extensive time skew.',
+    'items': ['Replication'],
+    'detail': """The time skew is over 365 days. If the time skew continues to
+increase eventually serious replication problems can occur.""",
+    'fix': """Avoid making changes to the system time, and make sure the clocks
+on all the replicas are correct."""
+}
+
 DSLOGNOTES0001 = {
     'dsle': 'DSLOGNOTES0001',
     'severity': 'Medium',
-- 
2.52.0

